src/azul/chalice.py

```diff
@@ -145,3 +161,160 @@ def _log_response(self, response):
+        self.route_specs[path].validate_params(self.current_request)
+
+    def _register_spec(self,
+                       path: str,
+                       path_spec: Optional[JSON],
+                       method_spec: Optional[JSON],
+                       methods: Iterable[str]):
+        """
+        Add a route's specifications to the specification object.
+        """
+        if path not in self.route_specs:
+            assert path not in self.specs['paths'], 'Only specify path_spec once per route path'
+            route_spec = RouteSpec(path, path_spec)
+            self.route_specs[path] = route_spec
+        else:
+            route_spec = self.route_specs[path]
+        if method_spec is not None:
+            for method in methods:
+                # OpenAPI requires HTTP method names be lower case
+                method = method.lower()
+                route_spec.set_route_spec(method, method_spec)
+        if len(route_spec.methods) > 0:
+            self.specs['paths'][path] = route_spec.to_dict()
+
+
+class RouteSpec:
+    """
+    Assists with managing details about API routes, provides access to request
+    validation.
+    """
+
+    def __init__(self, path, path_spec=None):
+        self.path: str = path
+        self.specs: MutableMapping[str, Dict[Any, Any]] = {}
+        self.validators: MutableMapping[str, Mapping[str, SpecValidator]] = {}
+
+        if path_spec:
+            self.path_spec = copy_json(path_spec)
+        else:
+            self.path_spec = {}
+
+    @property
+    def methods(self):
+        return [method for method in self.specs.keys()]
+
+    def get_method_spec(self, http_method) -> JSON:
+        """
+        Returns the method spec for a given API route
+        """
+        http_method = http_method.lower()
+        assert http_method in self.methods
+        return self.specs[http_method]
+
+    def get_spec_validators(self, http_method) -> Mapping[str, SpecValidator]:
+        """
+        Retuns a mapping of paramter names to their applicable SpecValidator.
+        """
+        http_method = http_method.lower()
+        assert http_method in self.methods
+        return self.validators[http_method]
+
+    def set_route_spec(self, http_method, method_spec):
+        """
+        Sets spec for a give route, creates request parameter SpecValidators
+        """
+        assert http_method.lower() not in self.methods, 'Only specify method_spec once per route path and method'
+        self.specs[http_method] = copy_json(method_spec)
+        self.validators[http_method] = self._create_spec_validators(http_method)
+
+    def extract_parameters(self, http_method: str) -> List[JSON]:
+        """
+        Returns a list of OpenAPI `parameters` that are applicable to a given API route
+        """
+        return [parameter for parameter in chain(self.path_spec.get('parameters', []),
+                                                 self.get_method_spec(http_method).get('parameters', []))]
```

Comment:
```suggestion
        return [*self.path_spec.get('parameters', []), *self.get_method_spec(http_method).get('parameters', []))]
```

It looks like you can remove the `chain` import now too.